File: scripts/build-og-selector.mjs
// Compose the /selector Open Graph image: a wall of the source channels'
// logos on the brand paper, closed by an acid bar. No font rendering here, so
// the words come from the og:title / og:description meta tags.
//
//   node scripts/build-og-selector.mjs   ->  img/selector/selector-og.png

import fs from 'node:fs';
import { execSync } from 'node:child_process';
import { channels } from '../selector-channels.mjs';

const W = 1200, H = 630;
const TILE = 132, GAP = 16, COLS = 7;
const BAR = 84;                       // acid bar across the bottom
const PAPER = [241, 238, 231], INK = [10, 10, 10], ACID = [255, 90, 54];

const slug = s => s.toLowerCase().normalize('NFKD').replace(/[̀-ͯ]/g, '')
  .replace(/[^a-z0-9]+/g, '-').replace(/^-|-$/g, '');

// read a png as an RGB pixel grid by way of sips + an uncompressed bmp
function readTile(png, size) {
  const tmp = '/tmp/_og_tile.bmp';
  execSync(`sips -z ${size} ${size} -s format bmp "${png}" --out ${tmp}`, { stdio: 'ignore' });
  const b = fs.readFileSync(tmp);
  const off = b.readUInt32LE(10), w = b.readInt32LE(18);
  let h = b.readInt32LE(22);
  const topDown = h < 0; h = Math.abs(h);
  const bpp = b.readUInt16LE(28) / 8;
  const row = Math.floor((bpp * w * 8 + 31) / 32) * 4;
  const px = new Uint8Array(w * h * 3);
  for (let y = 0; y < h; y++) {
    const sy = topDown ? y : h - 1 - y;
    for (let x = 0; x < w; x++) {
      const i = off + sy * row + x * bpp, o = (y * w + x) * 3;
      px[o] = b[i + 2]; px[o + 1] = b[i + 1]; px[o + 2] = b[i];
    }
  }
  return { px, w, h };
}

const canvas = new Uint8Array(W * H * 3);
const put = (x, y, c) => {
  if (x < 0 || y < 0 || x >= W || y >= H) return;
  const o = (y * W + x) * 3;
  canvas[o] = c[0]; canvas[o + 1] = c[1]; canvas[o + 2] = c[2];
};
const rect = (x0, y0, w, h, c) => {
  for (let y = y0; y < y0 + h; y++) for (let x = x0; x < x0 + w; x++) put(x, y, c);
};

rect(0, 0, W, H, PAPER);

const logos = [...new Set(channels.map(c => c.broadcaster))]
  .map(b => `img/selector/${slug(b)}.png`)
  .filter(p => fs.existsSync(p));

const rows = Math.ceil(logos.length / COLS);
const gridW = COLS * TILE + (COLS - 1) * GAP;
const gridH = rows * TILE + (rows - 1) * GAP;
const x0 = Math.round((W - gridW) / 2);
const y0 = Math.round((H - BAR - gridH) / 2);

logos.forEach((p, i) => {
  const cx = x0 + (i % COLS) * (TILE + GAP);
  const cy = y0 + Math.floor(i / COLS) * (TILE + GAP);
  const { px, w, h } = readTile(p, TILE);
  rect(cx - 3, cy - 3, TILE + 6, TILE + 6, INK);          // border
  for (let y = 0; y < h; y++) {
    for (let x = 0; x < w; x++) {
      const o = (y * w + x) * 3;
      put(cx + x, cy + y, [px[o], px[o + 1], px[o + 2]]);
    }
  }
});

rect(0, H - BAR, W, BAR, ACID);
rect(0, H - BAR - 4, W, 4, INK);

// 24bpp bottom-up BMP, then let sips make the png
const rowSize = Math.floor((24 * W + 31) / 32) * 4;
const body = Buffer.alloc(rowSize * H);
for (let y = 0; y < H; y++) {
  for (let x = 0; x < W; x++) {
    const s = ((H - 1 - y) * W + x) * 3, d = y * rowSize + x * 3;
    body[d] = canvas[s + 2]; body[d + 1] = canvas[s + 1]; body[d + 2] = canvas[s];
  }
}
const head = Buffer.alloc(54);
head.write('BM', 0);
head.writeUInt32LE(54 + body.length, 2);
head.writeUInt32LE(54, 10);
head.writeUInt32LE(40, 14);
head.writeInt32LE(W, 18);
head.writeInt32LE(H, 22);
head.writeUInt16LE(1, 26);
head.writeUInt16LE(24, 28);
head.writeUInt32LE(body.length, 34);
fs.writeFileSync('/tmp/_og.bmp', Buffer.concat([head, body]));

execSync('sips -s format png /tmp/_og.bmp --out img/selector/selector-og.png', { stdio: 'ignore' });
const kb = (fs.statSync('img/selector/selector-og.png').size / 1024).toFixed(0);
console.log(`img/selector/selector-og.png — ${W}x${H}, ${logos.length} logos, ${kb} KB`);

File: scripts/build-og-selector.py
"""Compose the /selector Open Graph image.

    python3 scripts/build-og-selector.py   ->  img/selector/selector-og.png

Replaces the Node version, which hand-wrote a BMP because it had no way to
draw text: a share card that says nothing reads as decoration in a feed, and
the title only arrives as a separate meta field. Pillow gives us real font
rendering, so the card can carry its own name and promise.

Space Mono is the site's mono face. It is fetched into a gitignored cache
rather than vendored, because the site serves it from Google Fonts and the
only thing this script produces, the PNG, is committed. Licence: OFL 1.1.
"""

import os
import re
import subprocess
import sys
import unicodedata

from PIL import Image, ImageDraw, ImageFont

W, H = 1200, 630
MARGIN = 72
BAR = 84                     # acid bar across the bottom
TILE, GAP = 90, 14
ROWS = (10, 9)               # logos per row

PAPER = (241, 238, 231)
INK = (10, 10, 10)
ACID = (255, 90, 54)
MUTED = (87, 83, 74)

TITLE = "The Selector"
CAPTION = "Press the button, pick a random DJ set"

FONT_CACHE = ".cache/fonts"
FONTS = {
    "bold": ("SpaceMono-Bold.ttf",
             "https://github.com/googlefonts/spacemono/raw/main/fonts/ttf/SpaceMono-Bold.ttf"),
    "regular": ("SpaceMono-Regular.ttf",
                "https://github.com/googlefonts/spacemono/raw/main/fonts/ttf/SpaceMono-Regular.ttf"),
}


def font(weight, size):
    name, url = FONTS[weight]
    path = os.path.join(FONT_CACHE, name)
    if not os.path.exists(path):
        os.makedirs(FONT_CACHE, exist_ok=True)
        print(f"fetching {name}")
        if subprocess.call(["curl", "-sfL", url, "-o", path]) != 0:
            sys.exit(f"could not fetch {name} from {url} (offline?)")
    return ImageFont.truetype(path, size)


def slug(value):
    value = unicodedata.normalize("NFKD", value.lower())
    value = "".join(c for c in value if not unicodedata.combining(c))
    return re.sub(r"^-|-$", "", re.sub(r"[^a-z0-9]+", "-", value))


def broadcasters():
    """The channel list lives in selector-channels.mjs; read the names out of it
    rather than keeping a second copy here that can drift."""
    source = open("selector-channels.mjs", encoding="utf-8").read()
    # names are single-quoted except where the name itself contains an
    # apostrophe ("L'Atelier de Musique"), so match on the opening quote
    names = [m[1] for m in re.findall(r"""broadcaster:\s*(['"])(.*?)\1""", source)]
    seen = []
    for name in names:
        if name not in seen:
            seen.append(name)
    return seen


def main():
    card = Image.new("RGB", (W, H), PAPER)
    draw = ImageDraw.Draw(card)

    draw.text((MARGIN, 74), TITLE, font=font("bold", 82), fill=INK)
    draw.text((MARGIN, 186), CAPTION, font=font("regular", 27), fill=MUTED)

    logos = [f"img/selector/{slug(b)}.png" for b in broadcasters()]
    logos = [p for p in logos if os.path.exists(p)]
    top = H - BAR - 52 - (len(ROWS) * TILE + (len(ROWS) - 1) * GAP)
    index = 0
    for row, count in enumerate(ROWS):
        y = top + row * (TILE + GAP)
        for column in range(count):
            if index >= len(logos):
                break
            tile = Image.open(logos[index]).convert("RGB").resize((TILE, TILE), Image.LANCZOS)
            card.paste(tile, (MARGIN + column * (TILE + GAP), y))
            index += 1

    draw.rectangle([0, H - BAR - 4, W, H - BAR], fill=INK)
    draw.rectangle([0, H - BAR, W, H], fill=ACID)

    out = "img/selector/selector-og.png"
    card.save(out, optimize=True)
    print(f"{out} - {W}x{H}, {index} logos, {os.path.getsize(out) // 1024} KB")


if __name__ == "__main__":
    main()
